Wip/sd 1028 2

cryogenian opened the pull request:
Open for discussion, ideas etc. 

I haven't implemented peeking from items (i.e.`move/rename` button click)

garyb commented on wip/notes:
```diff
@@ -0,0 +1,33 @@
+1.2. Names for terms. I suggest `comp` and `initialState`
+1.3. Name for coproducts: `ChildQuery`, `ChildState`, `ChildSlot`
+2. `Query`, `State` should be part of component module. `Slot` should be part of
+parent module.
```
Agree with all the above :+1:

garyb replied:
Although we'd been breaking out the component / `Query` / `State` into their own modules, like: https://github.com/beckyconning/slamdata-0.2.3/tree/86968d9f18387daf32bcd98189b69d927cac7000/src/UI/Notebook/Menu/Submenu which is helpful I think.

cryogenian replied:
`State` or `Model`? I'd prefer `State`. 
And reexport `MyComponent.State`, `MyComponent.Query` in `MyComponent`

garyb replied:
`State` works for me, sure. /cc @beckyconning

garyb commented on wip/notes:
```diff
@@ -0,0 +1,33 @@
+send `ItemAdd item` in `traverse` on every `Api.children`. I have many collections
+in my local mongodb instance. And after update it naturally freeze all laptop. It's
+not a problem, because after switching `ItemAdd Item -> ItemsAdd (Array Item)` it
+became smooth again; but I think it's worth to point it.
```
Yeah, that's definitely a downside to making lots of queries - it's because each change to the state forces a re-render, and also queries are processed asynchronously thanks to `Aff`, so there's quite a lot going on if you `traverse` a list of children to make queries.

garyb commented on wip/notes:
```diff
@@ -0,0 +1,33 @@
+1.3. Name for coproducts: `ChildQuery`, `ChildState`, `ChildSlot`
+2. `Query`, `State` should be part of component module. `Slot` should be part of
+parent module.
+3. Where are we going to put html snippets? I.e. `row`, `modalHeader`
```
@beckyconning might have some thoughts on that?

garyb commented on wip/notes:
```diff
@@ -0,0 +1,33 @@
+4. I'd modeled `g` as `Aff FileSystemEffects` and then created `Target` type synonim
+to simplify types and provide some abstractions. I suggest that we can remove type
+parameters by using concrete type synonims. After `SlamF` will be ready we can simply
+make `type Target = SlamF`.
```
Sounds good - we could just call it `Slam` or `SlamF` or something like that now perhaps, as that's less ambiguous than `Target`.

garyb commented on wip/notes:
```diff
@@ -0,0 +1,33 @@
+6. I suggest rename and put all arising model modules to `Data` package.
+7. It would be great if we will make some list of rules when we extract datatype
+and functions working with it to separate package.
+8. Shall we use lenses? Will we switch to `purescript-profunctor-lenses`?
```
To both of these: I'm not sure. I'd hope we don't need them too much now, since a lot of our previous use of them should be taken care of by the component plumbing - but if there are cases where they'd help (like navigating nasty structures in `peek`) then sure, why not? :smile:

I think `profunctor-lenses` is imperfect due to not having a story for indexed traversals, but I'm not sure how problematic that is for our needs.

garyb replied:
Oh, I see what you mean about lenses now - given that we have `newtype`'d `State` values, etc. I think it does still make sense to use them in that case, if we can't use synonyms for these values. I assume we newtyped them so we could provide Argonaut instances.

garyb commented on wip/notes:
````diff
@@ -0,0 +1,33 @@
+type Algebra =
+  Free (HalogenF State Query
+        (QueryF State ChildState Query ChildQuery Target ChildSlot))
+```
````
The `ParentDSL` synonym I just added should help with this, these can be added to Halogen as something like:

`liftAff'' :: forall eff a s f g. (MonadAff eff g, Functor g) => Aff eff a -> ParentDSL s s' f f' g p a`

garyb commented on wip/src/Control/UI/Browser.purs:
```diff
@@ -0,0 +1,16 @@
+foreign import select :: forall e. HTMLElement -> Eff (dom :: DOM |e) Unit
+foreign import newTab :: forall e. String -> Eff (dom :: DOM |e) Unit
+foreign import clearValue :: forall e. HTMLElement -> Eff (dom :: DOM|e) Unit
+foreign import reload :: forall e. Eff (dom :: DOM|e) Unit
```
Can we not use `purescript-dom` for a lot of this stuff? If not we should probably enhance it so we can and rely on that rather than doing loads of FFI stuff ourselves here.

cryogenian replied:
Do you mean _all_ this functions? Or just `select` and `clearValue`?

garyb replied:
I'm not sure :smile: - `replaceLocation`, `setLocation`, `reload` - things that are just an FFI call directly to the underlying API should probably be provided by `purescript-dom`, but perhaps not `clearValue`, as that's kinda a utility to `setValue ""` or something.

I don't think it's important for now, it could be a longer term goal. I just think that ideally we shouldn't need any direct FFI usage in `slamdata` itself.

garyb commented on wip/notes:
```diff
@@ -0,0 +1,33 @@
+in my local mongodb instance. And after update it naturally freeze all laptop. It's
+not a problem, because after switching `ItemAdd Item -> ItemsAdd (Array Item)` it
+became smooth again; but I think it's worth to point it.
+6. I suggest rename and put all arising model modules to `Data` package.
```
I'm not so sure about this one, I think there is benfit to differentiating `Model` from `Data` - I think of `Data.X` being for generic data structures, so something like `Data.Resouce` doesn't seem quite right.

garyb commented:
Ok, so I've not thoroughly reviewed all 7600+ lines of code :laughing:, but I think on the whole this is looking good - definitely seems like everything is being used the way it's "supposed" to be (I mean, we haven't really figured that out yet). There's some stuff we've inherited from the previous slamdata that could potentially be improved but I guess the main priority is to get everything working first!

Interested to see @jonsterling and @beckyconning's thoughts too.

jonsterling commented:
Likewise, I'm pleased by what I've looked at so far! I'll need a bit more time to review this in more detail. Thanks for all the hard work, @cryogenian!

beckyconning commented:
Cool stuff! This will take a while : )

cryogenian commented:
@jonsterling @garyb @beckyconning It's ready for review.
Travis fails because new psc build takes more then 3Gb :disappointed: 
+ Moved `Data.Resource` and other to `Model` package. Renamed `Foo.Utils` to `Utils.Foo`
+ Removed code that lives in `halogen-bootstrap`
+ File tests pass
+ Decomposed big components: `FileSystem`, `FileSystem.Dialog.Mount`, `FileSystem.Dialog.Download`
+ Removed model of notebook. I use to put `"""{"type": "new"}"""` during creating notebook, it's supposed
to decode this as new notebook domain model in notebook component. This removes adhesion between 
filesystem component and notebook component. It would be awesome if we could make notebook and 
file completely separate projects imo or at least share just very basic things like `Sort`. 
+ Common render functions from i.e. `Foo.Bar` and `Foo.Baz` moved to `Foo.Render`. 
+ Removed `qualified`s

I've used following rules to decompose components
+ don't decompose if component have less than 300 lines. It's easy enough to read even if it has children (because there is a lot of boilerplate: lenses, children etc)
+ `eval` and `peek` are left in main module. They can't be used by component that share same algebra. 
+ state extracted to `Component.State` 
+ type synonims and children machinery extracted to `Component.Install`
+ rendering functions extracted to `Component.Render` if render function have no dependency on `Install`. If render function depends on `Install` it can't be used in other component with same algebra.

The idea is to extract only those part of component that could be useful in creation of other components. 

I suggest 
+ to make branch in _slamdata/slamdata_ and make new version prs there. I.e. `2.3` or `new-halogen` or even `development`. 
+ do not remove `src` from that branch until we completely port project. It's usefull for pulling asap fixes from master.
+ continue to work in `wip` folder

jonsterling commented:
@cryogenian Can you try rebasing this on top of master? I've switched the Travis environment to something that should increase available RAM to 7.5Gb.

cryogenian commented:
Oh... Travis can build project. But no tests pass in its environment. I've checked and seems that performance and virtual-dom rerender changed and almost all `sequence` need `tryRepeatedlyTo`.

Really annoying since tests work correct on my machine. I'll reopen pr to `new-halogen` (if there is no objection with this name) branch with squashed commits on Monday (I hope).

jonsterling commented:
@cryogenian OK, thanks!

Yes, it's really annoying that changes in performance are observed by our tests... :frowning:
